packages/studio-be/src/studio/nlu/application/model-state/model-state-repo.ts

```diff
@@ -0,0 +1,92 @@
+import { TrainingStatus } from '@botpress/nlu-client'
+import * as sdk from 'botpress/sdk'
+import { StatusType } from './typings'
+
+export interface ModelStatePrimaryKey {
+  botId: string
+  language: string
+  status_type: StatusType
+}
+
+export interface ModelStateRow extends ModelStatePrimaryKey {
+  modelId: string
+  definitionHash: string
+  status: TrainingStatus
+  progress: number
+}
+
+export interface IModelStateRepository {
+  get(key: ModelStatePrimaryKey): Promise<ModelStateRow | undefined>
+  has(key: ModelStatePrimaryKey): Promise<boolean>
+  upsert(model: ModelStateRow): Promise<void>
+  insert(model: ModelStateRow): Promise<void>
+  update(model: Partial<ModelStateRow>): Promise<void>
+  del(key: ModelStatePrimaryKey): Promise<void>
+  query(query: Partial<ModelStateRow>): Promise<ModelStateRow[]>
+}
+
+export class DbModelStateRepository implements IModelStateRepository {
+  private _tableName = 'model_state'
+
+  constructor(private _db: sdk.KnexExtended) {}
+
+  public async initialize() {
+    await this._db.createTableIfNotExists(this._tableName, table => {
+      table.string('botId').notNullable()
+      table.string('language').notNullable()
+      table.string('status_type').notNullable()
+      table.string('modelId').notNullable()
+      table.string('definitionHash').notNullable()
+      table.string('status').notNullable()
+      table.float('progress').notNullable()
+      table.primary(['botId', 'language', 'status_type'])
+    })
+  }
```

Comment:
The studio is meant to be executed on a user's machine and it shouldn't require access to the database. The only reason we keep the database logic in the studio backend is so we can ship it right now and it work seamlessly for users in production.

It should certainly not create a database schema

Reply:
This information needs to be shared between different botpress instances... If not it won't work.

It's either in the kvs, the database or the ghost. I just thought the database was better suited

Reply:
I suggest using a file on the ghost to keep track of the progress. When the studio will run as a standalone, the user will build his bot locally, train it, and export the .tgz archive which will contain the trained model. Building it this way is future-proof for when we will stop shipping the studio with the rest of the product.

The ghost is future-proof, since when we will just ship the studio as  a standalone, we can simply remove the database logic and it will use the disk driver only